PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
 Code related to processing of import hooks.
 """
 
-import glob
+import glob, weakref
 import os.path
 
 from .. import log as logging
-from .utils import format_binaries_and_datas
-from ..compat import expand_path
-from ..compat import importlib_load_source
+from ..compat import (
+    expand_path, importlib_load_source, FileNotFoundError, UserDict)
 from .imphookapi import PostGraphAPI
+from .utils import format_binaries_and_datas
 
 logger = logging.getLogger(__name__)
 
 
+class ModuleHookCache(UserDict):
+    """
+    Cache of lazily loadable hook script objects.
+
+    This cache is implemented as a `dict` subclass mapping from the
+    fully-qualified names of all modules with at least one hook script to lists
+    of `ModuleHook` instances encapsulating these scripts. As a `dict` subclass,
+    all cached module names and hook scripts are accessible via standard
+    dictionary operations.
+
+    Attributes
+    ----------
+    module_graph : ModuleGraph
+        Current module graph.
+    _hook_module_name_prefix : str
+        String prefixing the names of all in-memory modules lazily loaded from
+        cached hook scripts. See also the `hook_module_name_prefix` parameter
+        passed to the `ModuleHook.__init__()` method.
+    """
+
+    _cache_id_next = 0
+    """
+    0-based identifier unique to the next `ModuleHookCache` to be instantiated.
+
+    This identifier is incremented on each instantiation of a new
+    `ModuleHookCache` to isolate in-memory modules of lazily loaded hook script
+    code objects in that cache to the same cache-specific namespace, preventing
+    edge-case collisions with existing in-memory modules in other caches.
+    """
+
+    def __init__(self, module_graph):
+        """
+        Initialize this cache.
+
+        Parameters
+        ----------
+        module_graph : ModuleGraph
+            Current module graph.
+        """
+
+        # Curiously, calling "super(UserDict, self).__init__()" here instead
+        # subsequently results in exceptions resembling:
+        #     AttributeError: 'ModuleHookCache' object has no attribute 'data'
+        UserDict.__init__(self)
```

Comment:
Your first intuition was right, actually. This _should_ be using `super`.

I initially thought the `UserDict` class was simply b0rked. Only after pushing that comment did I realize (with the sort of growing horror that gnaws at your gut) the truth: I'd originally refactored this code from the older `HooksCache` class, **which is subtly broken:**

```
class HooksCache(dict):
    def __init__(self, hooks_dir):
        # This doesn't actually do anything. This should instead be:
        #super(HooksCache, self).__init__()
        super(dict, self).__init__()
```

I now know: never trust the lies of old code.